lib/jwe.js

```diff
@@ -0,0 +1,511 @@
+var createBuffer = (data, encoding) => {
+  if (!Buffer.isBuffer(data) && typeof data !== 'string' && typeof data !== 'number')
+    throw new Error('in createBuffer, data must be a buffer, string or number');
+
+  if (process.version >= 'v6') {
+    if (typeof data === 'string')
+      return Buffer.from(data, encoding);
+    else if (typeof data === 'number')
+      return Buffer.alloc(data);
+    else
+      return Buffer.from(data);
+  } else {
+    if (typeof data === 'string')
+      return new Buffer(data, encoding);
+    else
+      return new Buffer(data);
+  } 
+};
+
+/**
+ * xor of two number in buffer format
+ *
+ * @param{Buffer}  a: first number
+ * @param{Buffer}  b: second number
+ * @return{Buffer}  a^b
+ */
+var xor = (a, b) => {
+  var la = a.length;
+  var lb = b.length;
+  var c1, c2;
+  if (la > lb) {
+    c1 = a; c2 =b;
+  } else {
+    c2 = a; c1 = b;
+  }
+  var c = createBuffer(c1);
+  for (var i = 1; i <= c2.length; i++)
+    c[c1.length-i] = c[c1.length-i] ^ c2[c2.length-i];
+  return c;
+};
+
+/******************************************************************************
+ * AES key unwrap algorithms
+ *****************************************************************************/
+
+/**
+ * Unwrap wrapped_cek using AES with kek
+ *
+ * @param{String}  algorithm: algorithm
+ * @param{Buffer}  wrapped_cek: wrapped key
+ * @param{Buffer}  kek: key encryption key
+ * @return{Buffer}  cek, the unwrapped key
+ */
+var AESKeyUnWrap = (algorithm, wrapped_cek, kek) => {
+
+  /****************************************************************************
+   * Inputs: CipherText, (n+1) 64-bit values {C0, C1, ..., Cn}, and 
+   *         Key, K (the KEK)
+   * Outputs: Plaintext, n 64-bit values {P0, P1, K, Pn}
+   ***************************************************************************/
+  var C = wrapped_cek;
+  var n = C.length/8-1;
+  var K = kek;
+
+  /****************************************************************************
+   * 1) Initialize variables
+   *    Set A = C[0]
+   *    For i = 1 to b
+   *      R[i] = C[i]
+   ***************************************************************************/
+  var A = C.slice(0,8);
+  
+  var R = [createBuffer(1)];
+  for (var i = 1; i <= n; i++)
+    R.push(C.slice(8*i, 8*i+8));
+
+  /****************************************************************************
+   * 2) compute intermediate values
+   *    For j = 5 to 0
+   *      For i = n to 1
+   *        B = AES-1(K, (A^t) | R[i]) where t = n*j+i
+   *        A = MSB(64, B)
+   *        R[i] = LSB(64, B)
+   ***************************************************************************/
+  for(var j=5; j >= 0; j--) {
+    for(var i=n; i >= 1; i--) {
+      // turn t = n*j+i into buffer
+      var str = (n*j+i).toString(16);
+      if (str.length %2 !== 0)
+        str = '0' + str;
+      var t = createBuffer(str, 'hex');
+
+      // B = AES-1(K, (A^t) | R[i])
+      var aes = crypto.createDecipheriv(algorithm, K, '');
```

Comment:
crypto.createDecipheriv will throw if there is any error. We catch all the key decryption exceptions in decryptCEKHelper function.